<a href="https://colab.research.google.com/github/Ayanlola2002/DATA-SCIENCE-PROJECTS/blob/master/Data_Wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [53]:
#mounting Gdrive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [54]:
import pandas as pd
import numpy as np
from datetime import date, datetime
import holidays

In [55]:
df_5min=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/data_five_min_id_47803.csv')
df_hourly=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/data_hourly_id_47803.csv')
df_daily=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/data_daily_id_47803.csv')
df_monthly=pd.read_csv('/content/drive/MyDrive/NEEDENERGY/data_monthly_id_47803.csv')

In [56]:
#numericalcol = [f for f in df_5min.columns if df_5min.dtypes[f] != 'object']


In [57]:
#df_5min[numericalcol]
#miss=df_5min[numericalcol].columns[df_5min[numericalcol].isnull().any()].tolist()

In [71]:
##Wrangling Helper Functions to Use
"""
This function do the reading of data file from file path(i.e drive)
"""
def do_read_data(datafile, dateCol):
    ''' Read the excel file'''
    df = pd.read_excel(datafile, index_col= dateCol, parse_dates=True)
    return df

"""
This function Shows Information of the Data file
Like The Header columns, Dimension of Data
"""

def do_data_information(datafile, info = False, head = True, shape = True):
    ''' Look at the info of the data'''
    
    if not isinstance(datafile, pd.DataFrame):
        df = read_data(datafile)
    else:
        df = datafile
    
    ## Information of data
    if info:
        df.info()
        
    print('--------------------------------------------------------------------')
    print('First rows')
    # Show first rows
    if head:
        print(df.head())
        
    print('--------------------------------------------------------------------')
    print('Dimension of the data')
    ## Dimension of data
    if shape:
        print(f'This data has {df.shape[0]} rows and {df.shape[1]} variables') 
        
###This function returns the number of distinct values from any columns       
def do_unique_values(df, col):
    ''' Find the unique values of a columns'''
    return df[col].nunique()


###This function statistically describe the data values
def do_description(df, col = None):
    ''' Give the statistic summary'''
    
    ## Summary statistic
    if col is not None:
        stat = df[col].describe()
    else:
        stat = df.describe()

    ## Change the name of the 50% index to median
    idx = stat.index.tolist()
    idx[5] = 'median'
    stat.index = idx
    return stat

###This function finds % Missing values
def check_missing_value(data):
    ## percentage of missing values
    n = data.isnull().sum().sort_values(ascending=False)/ len(data) * 100
    return n

def check_missing_values_table(df):
  mis_val = df.isnull().sum()
  mis_val_percent = 100 * df.isnull().sum() / len(df)
  mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
  mis_val_table_ren_columns = mis_val_table.rename(
  columns = {0 : 'Missing Values', 1 : '% of Total Values'})
  mis_val_table_ren_columns = mis_val_table_ren_columns[
      mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
  '% of Total Values', ascending=False).round(1)
  print ("Your selected dataframe has " + str(df.shape[1]) + " columns.\n"      
      "There are " + str(mis_val_table_ren_columns.shape[0]) +
        " columns that have missing values.")
  return mis_val_table_ren_columns


###This function finds the amount of each values in a dataset
def do_valueCount(data, col, perc = False):
    ''' calculate the proportion in fraction or percentage'''
    if not perc:
      return data[col].value_counts()
    else:
      return data[col].value_counts(normalize = True).mul(100).round(1).astype(str) + '%'    
### This function count number of unique values for all columns in a dataset
def do_nunique_value(df):
  print('The Numver of unique values per columns are as follows')
  print(df.agg(['nunique']).T)


###For date time conversion start here   
def do_date_col_convertion(df, column_name):
  df[column_name]=pd.to_datetime(df[column_name])
  return df[column_name]

def do_time_col_convertion(df, column_name):
    return pd.to_timedelta(df[column_name])


def do_datepart_extraction(data):
  data['date'] = pd.to_datetime(data['date'])
  data['year'] = data['date'].dt.year
  data['quarter']=data['date'].dt.quarter
  data['month'] = data['date'].dt.month
  data['weekday'] = data['date'].dt.weekday
  data['day_name'] = data['date'].dt.day_name()
  data['dayofyear'] = data['date'].dt.dayofyear
  data['day'] =data['date'].dt.day
  data['hour'] = data['date'].dt.hour
  data['minute'] =data['date'].dt.minute
  data['Season']=data['Month'].apply(do_year_season)
  data[Day_Season]=data['hour'].apply(do_day_season)
  return data

'''def do_datepart_extraction(data, col):
    # Create year, month, month name and day from a datetime column
    #data[col] = convert(data, col)
    
    data['Year'] = pd.DatetimeIndex(data[col]).year
    data['Month'] = pd.DatetimeIndex(data[col]).month
    data['Day'] = pd.DatetimeIndex(data[col]).day
    data['weekday'] = pd.DatetimeIndex(data[col]).weekday
    data['weekday_name'] = pd.DatetimeIndex(data[col]).day_name()
    data['Hour'] = pd.DatetimeIndex(data[col]).hour
    data['Quarter'].pd.DatetimeIndex(data[col]).Quarter

    return data
'''    

def do_year_season(data):
    ''' create season'''
    if data['month'] in (1, 2, 12):
        return "Winter"
    elif data['month'] in (3, 4, 5):
        return "Spring"
    elif data['month'] in (6, 7, 8):
        return "Summer"
    elif data['month'] in (9, 10, 11):
        return "Fall"
    else:
        return "Other"

def do_day_season(data):
  if data['hour'] >=0 and data['Hour'] < 5:
    return "Midnight"
  elif data['hour'] >=5 and data['Hour'] < 12:
    return "Morning"
  elif data['hour'] >=12 and data['Hour'] < 17:
    return "Afternoon"
  elif data['hour'] >=17 and data['Hour'] < 19:
    return "Evening"
  else:
    return "Night"


def get_holidays(startYear = 2018, endYear = 2025, countryCode = 'ZA'):
  """
    Takes in a start and end date, and start and end year.
    
    Produces a dataframe with a daily date and columns:
    holiday - 'Y' for holiday
    holidayName - name of the holiday if holiday is 'Y'
    
    Returns a dataframe
  """
  holidayDict = {}
  for i in range(startYear, endYear):
    for date, name in sorted(holidays.CountryHoliday(countryCode,years=[i]).items()):
      holidayDict[date] = name
      holiday_df = pd.DataFrame(list(holidayDict.items()),columns = ['day','holidayName'])
      holiday_df['day'] = pd.to_datetime(holiday_df['day'], format='%Y-%M-%d').dt.date
    return holiday_df
    

    
def get_days(start = '1/1/2018',startYear = 2018, end = '31/12/2025',endYear = 2025, countryCode = 'ZA'):
    """
    Takes in a start and end date, and start and end year.
    
    Produces a dataframe with a daily date and columns:
    weekend - 'Y' for weendend and 'N' for workday
    dayOfweek - numerical day of the week identifier 0 for monday
    weekNum - numerical number of the week 
    holiday - 'Y' for holiday
    holidayName - name of the holiday if holiday is 'Y'
    
    Returns a dataframe
    """
    
    #generate the range of daily dates
    dates = pd.date_range(start = start, end = end)
    date_df = pd.DataFrame(dates, columns = ['day'])
    date_df['day'] = pd.to_datetime(date_df['day'], format='%Y-%M-%d')
    country_holidays = get_holidays(startYear = startYear, endYear = endYear, countryCode = countryCode)
    
    date_df['dayName'] = pd.DatetimeIndex(date_df['day']).day_name()
    date_df['dayOfWeek'] = date_df['day'].dt.dayofweek
    date_df['weekend'] = date_df['dayOfWeek'].apply(lambda x: 'Y' if x>4 else 'N')
    date_df['weekNum'] = date_df['day'].dt.week
    date_df['holiday'] = date_df['day'].apply(lambda x: 'Y' if x in country_holidays['day'].values else 'N')
    date_df['day'] = date_df['day'].dt.date
    date_df = date_df.merge(country_holidays, on='day', how='left', indicator=False)
    
    date_df.to_csv(f'../data/public_holidays_weekends.csv', index=False)
    return date_df


def Impute_nan(df):
  numericalcol = [f for f in df_5min.columns if df_5min.dtypes[f] != 'object']
  missing_col=df_5min[numericalcol].columns[df_5min[numericalcol].isnull().any()].tolist()
  #Technique 1: Using mean to impute the missing values
  for i in missing_col:
    df.loc[df.loc[:,i].isnull(),i]=df.loc[:,i].mean()

def Impute_na(df):
  numericalcol = [f for f in df_5min.columns if df_5min.dtypes[f] != 'object']
  missing_col=df_5min[numericalcol].columns[df_5min[numericalcol].isnull().any()].tolist()
  #Technique 1: Using mean to impute the missing values
  for i in missing_col:
     df[i].fillna(mean,inplace=True)



#df['holidays']=df['date'].apply(lambda(x: 'Hols' if x in get_holidays()['day'].values else 'Non-Hols'))



In [61]:
#Data Description and Information for df_5min
print(do_data_information(df_5min))
print(do_data_information(df_hourly))
print(do_data_information(df_daily))
print(do_data_information(df_monthly))

--------------------------------------------------------------------
First rows
                  date  ...                                   voltageHarmonics
0  2020-03-15 18:15:00  ...  [[], [], [], [], [], [], [], [], [], [], [], [...
1  2020-03-15 18:20:00  ...  [[], [], [], [], [], [], [], [], [], [], [], [...
2  2020-03-15 18:25:00  ...  [[], [], [], [], [], [], [], [], [], [], [], [...
3  2020-03-15 18:30:00  ...  [[], [], [], [], [], [], [], [], [], [], [], [...
4  2020-03-15 18:35:00  ...  [[], [], [], [], [], [], [], [], [], [], [], [...

[5 rows x 15 columns]
--------------------------------------------------------------------
Dimension of the data
This data has 95649 rows and 15 variables
None
--------------------------------------------------------------------
First rows
                  date  ...                                   voltageHarmonics
0  2020-03-15 18:00:00  ...  [[], [], [], [], [], [], [], [], [], [], [], [...
1  2020-03-15 19:00:00  ...  [[], [], [], [], [

Data Quality Information:
df_5min : The data has 95649 rows and 15 features

In [62]:
#Checking Missing Values by percentage
print(check_missing_values_table(df_5min))
print(check_missing_values_table(df_daily))
print(check_missing_values_table(df_hourly))
print(check_missing_values_table(df_monthly))


Your selected dataframe has 15 columns.
There are 0 columns that have missing values.
Empty DataFrame
Columns: [Missing Values, % of Total Values]
Index: []
Your selected dataframe has 15 columns.
There are 0 columns that have missing values.
Empty DataFrame
Columns: [Missing Values, % of Total Values]
Index: []
Your selected dataframe has 15 columns.
There are 0 columns that have missing values.
Empty DataFrame
Columns: [Missing Values, % of Total Values]
Index: []
Your selected dataframe has 15 columns.
There are 0 columns that have missing values.
Empty DataFrame
Columns: [Missing Values, % of Total Values]
Index: []


In [63]:
#Checking number of Unique values
print(do_nunique_value(df_5min))
print(do_nunique_value(df_hourly))
print(do_nunique_value(df_daily))
print(do_nunique_value(df_monthly))

The Numver of unique values per columns are as follows
                  nunique
date                95649
timestamp           95649
consumption         76377
solar               50268
alwaysOn              955
gridImport          66188
gridExport              1
selfConsumption         2
selfSufficiency      5594
active              95512
reactive            95595
voltages              389
phaseVoltages         389
currentHarmonics        1
voltageHarmonics        1
None
The Numver of unique values per columns are as follows
                  nunique
date                 7979
timestamp            7979
consumption          7963
solar                7876
alwaysOn              714
gridImport           7946
gridExport              1
selfConsumption         2
selfSufficiency      3092
active               7979
reactive             7979
voltages              280
phaseVoltages         280
currentHarmonics        1
voltageHarmonics        1
None
The Numver of unique values per columns are as f

In [72]:
#Feature Engineering
df_5min=do_datepart_extraction(df_5min)


KeyError: ignored